rj_vision_receiver/src/vision_receiver.cpp

```diff
@@ -0,0 +1,362 @@
+#include <rj_vision_receiver/vision_receiver.h>
+
+#include <boost/exception/diagnostic_information.hpp>
+#include <cmath>
+#include <rj_common/Field_Dimensions.hpp>
+#include <rj_common/multicast.hpp>
+#include <rj_constants/topic_names.hpp>
+#include <rj_utils/logging.hpp>
+#include <stdexcept>
+
+namespace vision_receiver {
+using boost::asio::ip::udp;
+
+VisionReceiver::VisionReceiver()
+    : Node{"vision_receiver"}, config_{this}, port_{-1}, _socket{_io_context} {
+    _recv_buffer.resize(65536);
+
+    // There should be at most four packets: one for each camera on each of two
+    // frames, assuming some clock skew between this
+    // computer and the vision computer.
+    // TODO(#1528): Don't hardcode the number of cameras.
+    _packets.reserve(4);
+
+    // Parameters
+    declare_parameter<int>("port", SharedVisionPortSinglePrimary);
+    get_parameter<int>("port", port_);
+
+    double hz{};
+    declare_parameter<double>("hz", 60.0);
+    get_parameter<double>("hz", hz);
+
+    setPort(port_);
+
+    auto period = std::chrono::duration<double>(1 / hz);
+
+    raw_packet_pub_ =
+        create_publisher<RawProtobufMsg>(topics::kRawProtobufPub, 10);
+    detection_frame_pub_ =
+        create_publisher<DetectionFrameMsg>(topics::kDetectionFramePub, 10);
+
+    // Create timer for callback
+    timer_ = this->create_wall_timer(period, [this]() { run(); });
+}
+
+void VisionReceiver::setPort(int port) {
+    // If the socket is already open, close it to cancel any pending
+    // operations before we reopen it on a new port.
+    if (_socket.is_open()) {
+        _socket.close();
+    }
+
+    // Open the socket and allow address reuse. We need this to allow
+    // multiple instances of soccer to listen on the same multicast port.
+    _socket.open(udp::v4());
+    _socket.set_option(udp::socket::reuse_address(true));
+
+    // Set up multicast.
+    if (!multicast_add_native(_socket.native_handle(),
+                              SharedVisionAddress.c_str())) {
+        EZ_ERROR("Multicast add failed");
+        return;
+    }
+
+    // Bind the socket.
+    boost::system::error_code bind_error;
+    _socket.bind(udp::endpoint(udp::v4(), port), bind_error);
+    if (static_cast<bool>(bind_error)) {
+        EZ_ERROR_STREAM(
+            "Vision port bind failed with error: " << bind_error.message());
+        return;
+    }
+
+    // Start receiving. Note that this only listens once; any later receive
+    // will call `startReceive` again to continue listening.
+    startReceive();
+}
+
+void VisionReceiver::run() {
+    // If we haven't received config from the config server yet, return
+    if (!config_.connected()) {
+        return;
+    }
+    static bool connected = false;
+    if (!connected) {
+        connected = true;
+        EZ_INFO("Connected!");
+    }
+
+    // Let boost::asio check for new packets
+    _io_context.poll();
+
+    // Process and clear new packets.
+    processNewPackets();
```

Comment:
I'm keeping these in two separate steps so that the we get the receive time as accurate as possible if we want to try and estimate clock skew things later on.